polaris/hub/storage.py

```diff
@@ -0,0 +1,439 @@
+from base64 import b64encode
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from contextlib import contextmanager
+from hashlib import md5
+from io import BytesIO
+from pathlib import Path
+from typing import Any, Generator, Literal, Mapping, Sequence, TypeAlias
+
+import boto3
+from authlib.integrations.httpx_client import OAuth2Client
+from authlib.oauth2 import OAuth2Error
+from authlib.oauth2.rfc6749 import OAuth2Token
+from botocore.exceptions import BotoCoreError, ClientError
+from typing_extensions import Self
+from zarr.context import Context
+from zarr.storage import Store
+
+from polaris.hub.oauth import HubStorageOAuth2Token, StoragePaths
+from polaris.utils.errors import PolarisHubError
+from polaris.utils.types import ArtifactUrn
+
+Scope: TypeAlias = Literal["read", "write"]
+
+
+class S3StoreException(Exception):
+    """
+    Base exception for S3Store.
+    """
+
+
+class S3StoreCredentialsExpiredException(S3StoreException):
+    """
+    Exception raised when the S3 credentials have expired.
+    """
+
+
+@contextmanager
+def handle_s3_errors():
+    """
+    Standardize error handling for S3 operations.
+    """
+    try:
+        yield
+    except ClientError as e:
+        error_code = e.response["Error"]["Code"]
+        if error_code == "ExpiredToken":
+            raise S3StoreCredentialsExpiredException(f"Error in S3Store: Credentials expired: {e}") from e
+        else:
+            raise S3StoreException(f"Error in S3Store: {e}") from e
+    except BotoCoreError as e:
+        raise S3StoreException(f"Error in S3Store: {e}") from e
+
+
+class S3Store(Store):
+    """
+     A Zarr store implementation using a S3 bucket as the backend storage.
+
+    It supports multipart uploads for large objects and handles S3-specific exceptions.
+    """
+
+    _erasable = False
+
+    def __init__(
+        self,
+        path: str,
+        access_key: str,
+        secret_key: str,
+        token: str,
+        endpoint_url: str,
+        part_size: int = 10 * 1024 * 1024,  # 10MB
+        content_type: str = "application/octet-stream",
+    ) -> None:
+        bucket_name, prefix = path.split("/", 1)
+
+        self.s3_client = boto3.client(
+            "s3",
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key,
+            aws_session_token=token,
+            endpoint_url=endpoint_url,
+        )
+        self.bucket_name = bucket_name
+        self.prefix = prefix
+        self.part_size = part_size
+        self.content_type = content_type
+
+    def _full_key(self, key: str) -> str:
+        """
+        Converts a relative key to the full bucket key.
+        """
+        return f"{self.prefix}/{key}"
+
+    def _multipart_upload(self, key: str, value: bytes) -> None:
```

Comment:
Nice 💅! One thing we may need to consider is adding a retry mechanism for `upload_part` in case there are temporary network issues or failures. This would help make the upload process more resilient, especially for large files.

Just out of curiosity, we're not using [boto3.s3.transfer](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3.html) here because it doesn't allow us to set the `ContentMD5` and collect `ETag` values manually, correct?

Reply:
That's a good question. I browsed over transfer's API and it didn't seem to allow for Content-MD5 check. But taking a close look, there's a `ChecksumAlgorithm` that can be passed as an extra argument to the transfer configuration.

I'll see if that can work for what we need.

Reply:
Report from my experiment: The `ChecksumAlgorithm` option maps ultimately to the `x-amz-checksum-algorithm` header, even when set to **md5**. This header is not supported by R2, unfortunately. So we can't use the Transfer API.